initramfs/init.sh

```diff
@@ -0,0 +1,202 @@
+
+init="/sbin/init"
+ro=""
+root=""
+rootflags=""
+rootfstype="auto"
+
+shutdown()
+{
+	while [ 1 ]; do
+		poweroff
+		echo "Please remove power to complete shutdown."
+		sleep 10s
+	done
+}
+
+restart()
+{
+	echo "Rebooting in 5 seconds ..."
+	sleep 5s
+	reboot
+	modprobe sunxi_wdt || true
+	watchdog -T 1 -t 60 -F /dev/watchdog
+	shutdown
+}
+
+rescue_shell()
+{
+	set +eu
+	${BB_BIN} echo -en "\n" \
+		"##################################################\n" \
+		"#                   ▄▄▄      ▄▄▄                 #\n" \
+		"#                   ███████▄▄███▄                #\n" \
+		"#                 ▄███  ▄▄   ▄▄ ▐█               #\n" \
+		"#              ▄██  ██ ▀▀▀   ▀▀▀▐█▄              #\n" \
+		"#              ███▄▄██  ▄▄▄▄▄▄▄ ▐██              #\n" \
+		"#               ██████ ▀       ▀ ▐█              #\n" \
+		"#            ▄████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█▄           #\n" \
+		"#          ████▐█ ▐████████████████▌▐██▄         #\n" \
+		"#        ▐███▌▄▄█ ▐████████████████▌▐█▄█         #\n" \
+		"#         ███████ ███████▀▀ ▀██████▌▐██▌         #\n" \
+		"#         ███████ █████████ ██▐████▌▐██▌         #\n" \
+		"#        ▄███████ ██████▄██████████▌▐█▀█         #\n" \
+		"#        ████  ▐█ ▐████████████████▌▐█ █▌        #\n" \
+		"#        ▀███ ███   ▀▀▀▀            ▐██▀         #\n" \
+		"#           ▀▀████▄▄████████████████▀            #\n" \
+		"#             ▐████████▀▀████▀▀▌  █▌             #\n" \
+		"#            ▄██████  █  ▐███▄▄█▄▄██             #\n" \
+		"#            ████████▀▀▀█▀████▀    ██            #\n" \
+		"#            █████████    ▐███     ██            #\n" \
+		"#              ▀███▄█▄▄▄▄▄███▀▀▀▀▀▀▀             #\n" \
+		"#                                                #\n" \
+		"# Starting busybox rescue and recovery shell.    #\n" \
+		"# Tip: type help<enter> for available commands.  #\n" \
+		"#                                                #\n" \
+		"##################################################\n"
+	${BB_BIN} sh
+}
+
+critical_error()
+{
+	echo "A critical error has occurred, entering recovery mode."
+	rescue_shell
+	shutdown
+}
+
+boot_root()
+{
+	echo "Mounting ${root}."
+	mount -t ${rootfstype} -o exec,suid,dev,noatime,$rootflags,$ro "${root}" "${ROOT_MOUNT}"
+	kernel_umount
+	echo "Starting linux on ${root} of type ${rootfstype} with init=${init}."
+	exec switch_root /mnt/root "${init}"
+}
+
+find_and_run_update()
+{
+	echo "Checking for updates ..."
+	for dev in ${UPDATE_DEVICES}; do
+		if [ ! -b "${dev}" ]; then
+			continue
+		fi
+
+		echo "Attempting to mount ${dev}."
+		if ! mount -t f2fs,ext4,vfat,auto -o exec,noatime "${dev}" "${UPDATE_MOUNT}"; then
+			continue
+		fi
+
+		if [ ! -x "${UPDATE_SCRIPT}" ]; then
+			umount "${dev}"
+			echo "No executable update '${UPDATE_SCRIPT}' found on ${dev}, trying next."
+			continue
+		fi
+
+		echo "Found update on ${dev}, executing update ${UPDATE_SCRIPT}."
+		if ! "${UPDATE_SCRIPT}"; then
+			umount "${dev}"
+			echo "Update failed!"
+			critical_error
```

Comment:
I would argue this line is wrong because of the following scenario:
Unbrick code on USB drive can then fail as /dev/mmcblk1p3 may be checked and fail before going to /dev/sda1.

Reply:
you are right! USB should be first!

Reply:
found some new stuff: booting with USB stick that had a update on an ext4 formated partition. (in previous tests vFat wasn't mounted) the USB drive wasn't checked at all!

[    1.367142] usb usb3: New USB device found, idVendor=1d6b, idProduct=0002
[    1.373934] usb usb3: New USB device strings: Mfr=3, Product=2, SerialNumber=1
[    1.381192] usb usb3: Product: MUSB HDRC host driver
[    1.386175] usb usb3: Manufacturer: Linux 4.14.32-ultimaker-61844-gb1ca4f281c2b musb-hcd
[    1.394266] usb usb3: SerialNumber: musb-hdrc.1.auto
[    1.399792] hub 3-0:1.0: USB hub found
[    1.403590] hub 3-0:1.0: 1 port detected
[    1.408340] sunxi-rtc 1c20d00.rtc: setting system clock to 2018-04-30 14:23:50 UTC (1525098230)
[    1.417515] vcc3v0: disabling
[    1.420499] vcc5v0: disabling
[    1.423484] ahci-5v: disabling
[    1.426602] ALSA device list:
[    1.429576]   No soundcards found.
[    1.435671] Freeing unused kernel memory: 2048K
Checking for updates ...
Attempting to mount /dev/mmcblk1p1.[    1.465175] F2FS-fs (mmcblk1p1): Magic Mismatch, valid(0xf2f52010) - read(0x2000)

[    1.473129] F2FS-fs (mmcblk1p1): Can't find valid F2FS filesystem in 1th superblock
[    1.481334] F2FS-fs (mmcblk1p1): Magic Mismatch, valid(0xf2f52010) - read(0x2005)
[    1.488846] F2FS-fs (mmcblk1p1): Can't find valid F2FS filesystem in 2th superblock
[    1.496581] F2FS-fs (mmcblk1p1): Magic Mismatch, valid(0xf2f52010) - read(0x2000)
[    1.504066] F2FS-fs (mmcblk1p1): Can't find valid F2FS filesystem in 1th superblock
[    1.511735] F2FS-fs (mmcblk1p1): Magic Mismatch, valid(0xf2f52010) - read(0x2005)
[    1.519238] F2FS-fs (mmcblk1p1): Can't find valid F2FS filesystem in 2th superblock
[    1.534663] EXT4-fs (mmcblk1p1): mounted filesystem with ordered data mode. Opts: (null)
No executable update '/mnt/update/um_update.sh' found on /dev/mmcblk1p1, trying next.
Attempting to mount /dev/mmcblk1p2.
[    1.604448] F2FS-fs (mmcblk1p2): Mounted with checkpoint version = bd0c
No executable update '/mnt/update/um_update.sh' found on /dev/mmcblk1p2, trying next.
Attempting to mount /dev/mmcblk1p3.
[    1.672304] F2FS-fs (mmcblk1p3): Mounted with checkpoint version = 16
No executable update '/mnt/update/um_update.sh' found on /dev/mmcblk1p3, trying next.
No updates found.